### ISSUE TITLE
[MCOMPILER-181] Exception on compilation are not properly reported

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-compiler-plugin#806]
**[Clebert Suconic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=clebertsuconic)** opened **[MCOMPILER-181](https://issues.apache.org/jira/browse/MCOMPILER-181?redirect=false)** and commented

I'm facing a NullPointerException on my project (HornetQ/master) when we update the maven-compiler to 2.5.1 or 2.5. It all works fine with 2.3.2 and as soon as we update it we start having these NPEs.

It's funny that's only happening on MacOS for us while all our Linux Envs are working fine. I'm not sure the root cause but it may be something that would eventually happen in other environments as well.

Steps to reproduce it:

I don't have an isolated issue, but you can reproduce on my project:
On a MacOS:

git clone git://github.com/hornetq/hornetq.git
edit the maven-compiler version to 2.5.1 on the main pom

and use mvn package

[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.592s
[INFO] Finished at: Fri Sep 14 09:22:23 CDT 2012
[INFO] Final Memory: 9M/81M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hornetq-commons: Compilation failure
[ERROR] error: java.lang.NullPointerException
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hornetq-commons: Compilation failure
error: java.lang.NullPointerException

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
    at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
    at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
    at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)

Caused by: org.apache.maven.plugin.CompilationFailureException: Compilation failure
error: java.lang.NullPointerException

    at org.apache.maven.plugin.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:729)
    at org.apache.maven.plugin.CompilerMojo.execute(CompilerMojo.java:128)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
    ... 19 more

[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

---

**Affects:** 2.5, 2.5.1

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

A bit weird the full stack trace is not returned which is bad

```
error: java.lang.NullPointerException
	at com.sun.codemodel.internal.JCodeModel.ref(JCodeModel.java:347)
	at org.jboss.logging.processor.model.ClassModel.generateModel(ClassModel.java:137)
	at org.jboss.logging.processor.model.ImplementationClassModel.generateModel(ImplementationClassModel.java:75)
	at org.jboss.logging.processor.model.MessageBundleImplementor.generateModel(MessageBundleImplementor.java:59)
	at org.jboss.logging.processor.model.ClassModel.create(ClassModel.java:104)
	at org.jboss.logging.processor.apt.ImplementationClassGenerator.processTypeElement(ImplementationClassGenerator.java:63)
	at org.jboss.logging.processor.apt.LoggingToolsProcessor.process(LoggingToolsProcessor.java:154)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:627)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:556)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:701)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:987)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:727)
	at com.sun.tools.javac.main.Main.compile(Main.java:353)
	at com.sun.tools.javac.main.Main.compile(Main.java:279)
	at com.sun.tools.javac.main.Main.compile(Main.java:270)
	at com.sun.tools.javac.Main.compile(Main.java:87)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess0(JavacCompiler.java:551)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess(JavacCompiler.java:526)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compile(JavacCompiler.java:167)
	at org.apache.maven.plugin.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:655)
	at org.apache.maven.plugin.CompilerMojo.execute(CompilerMojo.java:117)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:560)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:201)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
error: java.lang.NullPointerException
	at com.sun.codemodel.internal.JCodeModel.ref(JCodeModel.java:347)
	at org.jboss.logging.processor.model.ClassModel.generateModel(ClassModel.java:137)
	at org.jboss.logging.processor.model.ImplementationClassModel.generateModel(ImplementationClassModel.java:75)
	at org.jboss.logging.processor.model.MessageLoggerImplementor.generateModel(MessageLoggerImplementor.java:88)
	at org.jboss.logging.processor.model.ClassModel.create(ClassModel.java:104)
	at org.jboss.logging.processor.apt.ImplementationClassGenerator.processTypeElement(ImplementationClassGenerator.java:63)
	at org.jboss.logging.processor.apt.LoggingToolsProcessor.process(LoggingToolsProcessor.java:154)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:627)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:556)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:701)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:987)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:727)
	at com.sun.tools.javac.main.Main.compile(Main.java:353)
	at com.sun.tools.javac.main.Main.compile(Main.java:279)
	at com.sun.tools.javac.main.Main.compile(Main.java:270)
	at com.sun.tools.javac.Main.compile(Main.java:87)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess0(JavacCompiler.java:551)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compileInProcess(JavacCompiler.java:526)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.compile(JavacCompiler.java:167)
	at org.apache.maven.plugin.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:655)
	at org.apache.maven.plugin.CompilerMojo.execute(CompilerMojo.java:117)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:560)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:201)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
warning: Unclosed files for the types '[org.hornetq.utils.HornetQUtilBundle_$bundle, org.hornetq.utils.HornetQUtilLogger_$logger]'; these types will not undergo annotation processing
2 errors
1 warning
```

I will try to find the cause.

**Comment by jira-importer** [comment]
**[Clebert Suconic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=clebertsuconic)** commented

There's some code generation from jboss-logging.

on my project this is generated at hornetq-checkout-dir/hornetq-commons/target/generated-sources/annotations/org/hornetq/utils

Maybe that information will help you identify the root cause.

**Comment by jira-importer** [comment]
**[Clebert Suconic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=clebertsuconic)** commented

Hmmm... it's failing on the processor from jboss-logging ....

What difference could maven-compiler pluging have on that?

**Comment by jira-importer** [comment]
**[Clebert Suconic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=clebertsuconic)** commented

I was going to rename this JIRA to

"Exception on compilation are not properly reported"

Which will probably be an easy issue for you...

I already talked to the jboss-logging lead, this is a bug on their code.

If the exception was properly reported I wouldn't have bothered you guys...

But thanks a lot for the help

**Comment by jira-importer** [comment]
**[Clebert Suconic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=clebertsuconic)** commented

The root cause is a bug on the jboss-logging code, if you could rename this JIRA as "exceptions during compilation are not properly reported".

If there's any policy against renaming it... I can create a new one and close this one.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

renamed.
BTW the current plexus compiler api is poor for enhancement. Because here we need at least the full output.

```
List<CompilerError> compile( CompilerConfiguration configuration )
I would prefer 
CompilerResult compile( CompilerConfiguration configuration )
with
CompilerResult {
  boolean success;
  List<CompilerError> compilerErrors;
}
```

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

https://github.com/sonatype/plexus-compiler/pull/9 will fix that with using javax.tools compiler.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

fixed with using plexus-compiler 2.0 which can use javax.tools.* for compilation and have a better diagnostic
